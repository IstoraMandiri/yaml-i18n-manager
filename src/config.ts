import yaml from 'yaml';
import { promises } from 'fs';
import { resolve } from 'path';
import { log } from './utils';

const { writeFile, readFile, mkdir } = promises;

export const defaultConfig = {
  defaultLocale: 'en',
  contentDir: 'content',
  yamlDir: '_i18n',
  csvDir: 'csv',
  filterKeys: ['link', 'className', 'date', 'tags'],
  filterMarkdown: [],
};

export async function initConfig(argv: any) {
  const [command] = argv._;
  const location = resolve(process.env.PWD, argv.config);
  if (command === 'init') {
    try {
      const path = location.split('/').slice(0, -1).join('/');
      await mkdir(path, { recursive: true });
      await writeFile(location, yaml.stringify(defaultConfig));
      log(`Wrote config to ${argv.config}`, defaultConfig);
      return null;
    } catch (e) {
      throw e;
    }
  }
  let configData, config;
  try {
    configData = await readFile(location, 'utf8');
  } catch (e) {
    log(location);
    log(
      'Config file not found. Pass a config file with `--config` or initialize with `init` command.',
    );
    return null;
  }
  try {
    config = yaml.parse(configData);
  } catch (e) {
    log(`Could not parse the config file at ${location}`);
    return null;
  }
  return config;
}
