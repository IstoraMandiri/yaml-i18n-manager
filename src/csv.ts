import csv from 'async-csv';

const columns = [
  { key: 'dHash', header: 'dHash' },
  { key: 'path', header: 'Location' },
  { key: 'sig', header: 'Key' },
  { key: 'defaultValue', header: 'Default Language' },
  { key: 'value', header: 'Existing Translation' },
  { key: 'update', header: 'Updated Translation (or leave blank)' },
];

export default async function generate(data: TranslatedYaml[]) {
  return await csv.stringify(data, { header: true, columns });
}
