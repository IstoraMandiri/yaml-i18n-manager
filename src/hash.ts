import { createHash } from 'crypto';

export default function (str: string): string {
  return createHash('sha256').update(str, 'utf8').digest('base64').slice(0, 10);
}
