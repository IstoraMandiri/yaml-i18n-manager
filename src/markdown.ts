import yaml from 'yaml';
import fs from 'fs-extra';
import { resolve } from 'path';
import { Config } from './types';
import { scanDir } from './utils';
import { translateYaml } from './translate';

async function parseMarkdonwFile(file: string): Promise<{ header?: any; body: string }> {
  const data = await fs.readFile(file, 'utf8');
  let header, body;
  if (data.startsWith('---')) {
    const [_, _header, _body] = data.split('---');
    body = _body;
    header = yaml.parse(_header);
  } else {
    body = data;
  }
  return { header, body };
}

function normalize(n: string): string {
  const str = n.split('.').slice(0, -2).join('.');
  console.log(str);
  return str;
}

export async function generateMarkdown(locale: string, opts: Config) {
  const markdownFiles = (await scanDir(opts.contentDir))
    .filter(({ ext }) => ['md', 'mdx'].includes(ext))
    .filter(({ relative }) => !opts.filterMarkdown.find((p) => relative.startsWith(p)))
    .map((md) => ({ ...md, locale: md.name.split('.').slice(-2)[0] }))
    .filter(({ locale }) => [opts.defaultLocale, locale].includes(locale));
  const deduped = markdownFiles.filter(
    ({ relative, locale }) =>
      locale === opts.defaultLocale &&
      !markdownFiles.find(
        ({ relative: r2, locale: l2 }) => l2 !== locale && normalize(r2) === normalize(relative),
      ),
  );
  const contents = await Promise.all(
    deduped.map(async (file) => ({ ...file, data: await parseMarkdonwFile(file.path) })),
  );
  const normalized = contents.reduce((a, i): any => {
    const {
      data: { header, body },
      relative,
    } = i;
    const items = [{ relative, key: 'body', value: body }];
    Object.keys(header)
      .filter((k) => !opts.filterKeys.includes(k))
      .forEach((k) => {
        items.push({ relative, key: `header.${k}`, value: header[k] });
      });
    return [...a, ...items];
  }, []);

  const translated = await translateYaml(normalized, opts.defaultLocale, locale);
  const constructed = translated.reduce((o: any, { relative, key, value }: any) => {
    const [k, subKey] = key.split('.');
    const update = o[relative] || {};
    update[k] = subKey ? { ...update[k], [subKey]: value } : value;
    return { ...o, [relative]: update };
  }, {});
  // now write the files.
  await Promise.all(
    Object.keys(constructed).map(async (k) => {
      const newFileName = `${k.split('.').slice(0, -2).join('.')}.${locale}.${k.split('.').pop()}`;
      const { header, body } = constructed[k];
      const data = `---\n${yaml.stringify(header).trim()}\n---${body}`;
      await fs.writeFile(resolve(opts.contentDir, newFileName), data);
    }),
  );
}
