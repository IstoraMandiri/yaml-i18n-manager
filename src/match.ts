import { normalizeSig } from './utils';

export default function match(yamls: ExistingYaml[]): ExistingYaml[] {
  return yamls.map((dYaml) => {
    const matches = yamls.reduce((o, tYaml) => {
      if (tYaml.locale == dYaml.locale || normalizeSig(tYaml) !== normalizeSig(dYaml)) {
        return o;
      }
      return { [tYaml.locale]: tYaml };
    }, {});
    return { ...dYaml, matches };
  });
}
