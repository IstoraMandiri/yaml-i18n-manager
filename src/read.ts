import YAML from 'yaml';
import { resolve, relative } from 'path';
import { promises } from 'fs';
import { hash, K_SEP } from './utils';
const { readdir, readFile } = promises;

import type { ExistingYaml } from './types';

interface ParsedFile {
  absolutePath: string;
  relativePath: string;
  yaml: any[];
}

async function parseFile(targetDir: string, absolutePath: string): Promise<ParsedFile> {
  const data = await readFile(absolutePath, 'utf8');
  const yaml = await YAML.parse(data);
  return {
    absolutePath,
    relativePath: relative(targetDir, absolutePath),
    yaml,
  };
}

async function getFiles(root: string, dir?: string): Promise<ParsedFile[]> {
  const target = dir || root;
  const dirEntries = await readdir(target, { withFileTypes: true });
  const files = await Promise.all(
    dirEntries.map((entry): any => {
      const res = resolve(target, entry.name);
      if (entry.isDirectory()) {
        return getFiles(root, res);
      }
      if (entry.name.endsWith('.yaml') && !entry.name.includes('.collection.')) {
        return parseFile(root, res);
      }
    }),
  );
  return Array.prototype.concat(...files.filter((i) => i));
}

function flatten<T extends Record<string, any>>(
  object: T,
  path: string | null = null,
  separator = '__',
): T {
  return Object.keys(object).reduce((acc: T, key: string): T => {
    // always ignore `key` key
    // ignore non-strings
    if (key === 'key') {
      return acc;
    }
    let newPath = [path, key].filter(Boolean).join(separator);
    const i = key;
    const k = object?.[key].key;
    if (k) {
      newPath = [path, `[${i}:${k}]`].filter(Boolean).join(separator);
    }
    return typeof object?.[key] === 'object'
      ? { ...acc, ...flatten(object[key], newPath, separator) }
      : // filter non-strings
      typeof object[key] !== 'string'
      ? acc
      : { ...acc, [newPath]: object[key] };
  }, {} as T);
}

function normalize(files: ParsedFile[]): ExistingYaml[] {
  return files.reduce((a1, { yaml, relativePath: path }): any => {
    const [locale] = path.split('.').slice(-2);
    const flat = flatten(yaml) as { [k: string]: any };
    return [
      ...a1,
      ...Object.keys(flat).reduce(
        (a2, sig): any => [
          ...a2,
          {
            path,
            locale,
            sig,
            key: sig.split(K_SEP).slice(-1)[0],
            value: flat[sig],
            vHash: hash(flat[sig]),
          },
        ],
        [],
      ),
    ];
  }, []);
}

export default async function read(targetDir: string) {
  const files = await getFiles(targetDir);
  return normalize(files);
}
