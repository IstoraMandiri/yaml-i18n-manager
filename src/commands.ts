import { log } from './utils';
import { parse, generate as generateCsv } from './csv';
import eject from './eject';

import type { Config } from './types';
import generateYaml, { getYamlTranslations } from './yaml';
import { generateMarkdown } from './markdown';

const commands: any = {
  generate: async ({ locale }: { locale: string }, config: Config) => {
    if (locale === 'all') {
      for (const l of config.locales) {
        log(`Processing ${l}`);
        await commands.generate({ locale: l }, config);
      }
      log(`Completed translating ${config.locales.length} locales`);
      return;
    }
    // generate
    await generateMarkdown(locale, config);
    const translations = await getYamlTranslations(locale, config);
    const outputDir = `${config.outputDir || config.contentDir}/${config.yamlDir}`;
    const filePath = `${outputDir}/${locale}.yaml`;
    await generateYaml(translations, filePath);
    log(`Wrote ${translations.length} translations to ${filePath}`);
  },
  eject: async ({ locale }: { locale: string }, config: Config) => {
    const translations = await getYamlTranslations(locale, config);
    const outputDir = config.outputDir || config.contentDir;
    await eject(translations, outputDir);
    log(`Wrote ${translations.length} translations to ${outputDir}`);
  },
  export: async ({ locale }: { locale: string }, config: Config) => {
    const translations = await getYamlTranslations(locale, config);
    const filePath = `${config.csvDir}/${locale}.csv`;
    await generateCsv(translations, filePath);
    log(`Wrote to ${filePath}`);
  },
  import: async ({ csv }: { csv: string }, config: Config) => {
    const imports = await parse(csv);
    const outputDir = config.outputDir || config.contentDir;
    await eject(imports, outputDir);
    log(`Wrote ${imports.length} imports to ${outputDir}`);
  },
};

export default commands;
