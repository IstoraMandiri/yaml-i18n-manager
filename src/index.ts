#!/usr/bin/env node
import yargs from 'yargs';
import { defaultConfig, initConfig } from './config';
import commands from './commands';

function requireLocale(y: any) {
  y.positional('locale', {
    describe: 'Target language code (e.g. ja)',
    type: 'string',
  });
}

const argv: any = yargs
  .usage('Usage: $0 <command> [options]')
  .version('v')
  .alias('v', 'version')
  .help('h')
  .alias('h', 'help')
  .options({
    c: {
      alias: 'config',
      type: 'string',
      describe: 'Config file',
      default: `${defaultConfig.managementDir}/config.yaml`,
    },
  })
  .command('init', 'Generate Config File')
  .command('generate <locale>', 'Auto-translate content', requireLocale)
  .command('export <locale>', 'Export CSV for a locale', requireLocale)
  .command('import <csv>', 'Import CSV translations', (y: any) => {
    y.positional('csv', {
      describe: 'Relative CSV file path',
      type: 'string',
    });
  })
  .demandCommand(1, 'Please enter a command').argv;

(async () => {
  const config = await initConfig(argv);
  if (!config) {
    return;
  }
  const [command] = argv._;
  await commands[command](argv, config);
})();
