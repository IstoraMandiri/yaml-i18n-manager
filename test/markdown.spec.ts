import { generateMarkdown } from '../src/markdown';
import fs from 'fs-extra';
import { defaultConfig } from '../src/config';
import rimraf from 'rimraf';
import scan from './util/scan';

const contentSrc = './test/data/content-input';
const contentDir = './test/data/test-markdown';

const config = { ...defaultConfig, contentDir };

describe('commands', () => {
  beforeEach(async () => {
    rimraf.sync(contentDir);
    await fs.copy(contentSrc, contentDir);
  });
  afterEach(() => {
    rimraf.sync(contentDir);
  });
  describe('markdown', () => {
    it('translates markdown', async () => {
      await generateMarkdown('ja', config);
      expect(await scan(contentDir)).toMatchSnapshot();
    });
  });
});
