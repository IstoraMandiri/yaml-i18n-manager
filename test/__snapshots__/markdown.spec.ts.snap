// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands markdown translates markdown 1`] = `
Array [
  Object {
    "content": "# updated here
ui.global/footer: |
  Footer here
# English Here...
menu/hello: |
  This is a translated key
# original title here
sub/welcome/items__test__title: |
  This is the title!
",
    "fileName": "test/data/test-markdown/_i18n/ja.yaml",
  },
  Object {
    "content": "- key: home
  title: Home
  link: /
- key: about
  title: About
  link: /about
- key: news
  title: News
  link: /news
",
    "fileName": "test/data/test-markdown/menu.en.yaml",
  },
  Object {
    "content": "## I exist already
",
    "fileName": "test/data/test-markdown/sub/existing.en.md",
  },
  Object {
    "content": "## It wont be created
",
    "fileName": "test/data/test-markdown/sub/existing.ja.md",
  },
  Object {
    "content": "---
title: Test title
tags: ['a', 'b', 'c']
description: Description goes here
---

## I am some markdown. Please translate me!

Markdown is a lightweight markup language with plain-text-formatting syntax. Its design allows it to be converted to many output formats, but the original tool by the same name only supports HTML. [Markdown](https://en.wikipedia.org/wiki/Markdown) is often used to format readme files, for writing messages in online discussion forums, and to create rich text using a plain text editor.

Since the initial description of Markdown contained ambiguities and unanswered questions, the implementations that appeared over the years have subtle differences and many come with syntax extensions.
",
    "fileName": "test/data/test-markdown/sub/index.en.md",
  },
  Object {
    "content": "---
title: 試験タイトル
description: ここに説明が入ります
---

## 私はいくつかの値下げです。翻訳してください!

Markdownは、プレーンテキスト形式の構文を持つ軽量のマークアップ言語です。その設計により、多くの出力形式に変換できますが、同じ名前の元のツールはHTMLのみをサポートします。 [Markdown](https://en.wikipedia.org/wiki/Markdown)は、readmeファイルのフォーマット、オンラインディスカッションフォーラムでのメッセージの書き込み、プレーンテキストエディターを使用したリッチテキストの作成によく使用されます。

Markdownの最初の説明にはあいまいさと未回答の質問が含まれていたため、長年にわたって登場した実装には微妙な違いがあり、その多くに構文拡張が付属しています。
",
    "fileName": "test/data/test-markdown/sub/index.ja.md",
  },
  Object {
    "content": "intro: Welcome to this sub folder
description: This is a description that won't be translated
items:
  - key: test
    title: Testing
  - key: hello
    title: Hello Work
",
    "fileName": "test/data/test-markdown/sub/welcome.en.yaml",
  },
  Object {
    "content": "title: Title
description: This is a description
new: Some new key
body: |
  ## This is some markdown

  How is this?
",
    "fileName": "test/data/test-markdown/ui.global.en.yaml",
  },
]
`;
